test: add an easy way to run linters locally

jamesob opened the pull request:
Adds a Dockerfile configuration ~~(originally written mostly by fanquake)~~ that allows straightforward running of linters with compatible versions locally. This removes a ton of annoyance when trying to appease CI, because many of the linter versions are quite old and difficult to maintain locally.

I realize that people may not be thrilled to add more ancillary tooling to the repo, but I think this makes a lot of sense given the linter versions listed in this container configuration are dictated by this repo (within the CI configuration), so having these things live in two separate places is a recipe for version mismatches.

Eventually we can likely just use this container on CI directly to avoid any chance of inconsistencies between local dev experience and CI.

jamesob committed:
test: add an easy way to run linters locally

Adds a Dockerfile configuration
that allows straightforward running of linters with compatible versions
locally. This removes a ton of annoyance when trying to appease CI,
because many of the linter versions are quite old and difficult to
maintain locally.

I realize that people may not be thrilled to more ancillary tooling to
the repo, but I think this makes a lot of sense given the linter
versions listed in this container configuration are dictated by this
repo (within the CI configuration), so having these things live in
two separate places is a recipe for version mismatches.

Eventually we can likely just use this container on CI directly to avoid
any chance of inconsistencies between local dev experience and CI.

MarcoFalke commented on contrib/devtools/lint.dockerfile:
```diff
@@ -0,0 +1,49 @@
+ARG PYTHON=3.6.15
+ARG PYZMQ=24.0.1
+ARG SHELLCHECK=v0.8.0
+ARG VULTURE=2.6
```
You claim that

> live in two separate places is a recipe for version mismatches.

But in this pull you are duplicating the versions to two places in this repo. (As well as all other stuff)

maflcko reviewed (commented):
Not sure about duplicating the versions and everything else. As you say, this is a recipe for errors.

Also, the lint CI fails on this

jamesob commented:
Okay well I don't really care which way we go here, but the situation right now is really untenable because without this dockerfile it's basically not feasible to replicate linting locally. As far as I can tell, it either lives here or in a third party repo.

Do you have an alternate suggestion that will make it reasonable for developers to run the CI lints locally?

MarcoFalke commented:
I am not too fluent in docker, so I can't provide alternative suggestions. However, if you find a way to not duplicate everything, it should be more acceptable to merge?

jamesob commented:
I don't even care about docker, I just want a way to run linters locally. How do you do it?

MarcoFalke commented:
> How do you do it?

I don't run the linters. When I write a scripted diff I call the script-check directly; I know not to modify subtrees, so I don't need to lint for that; And for the python stuff I just fix it up after opening a PR.

Depending on how many features you want if you run locally, there are alternatives:
* If you only want to run the python stuff, Extracting the versions into a requirements.txt might be enough before calling lint-all.
* If you want to run more than that, it can be extracted into new scripts from the scripts in `./ci/lint/`. (Excluding the destructive stuff like `sed -i -e 's/bionic/jammy/g' /etc/apt/sources.list`, ...)
* If you want to run the exact CI env, I don't think it will be possible without docker.

MarcoFalke commented:
(Related merge commit: 01ec5308bf616740c804247043046b23b483df5c)

jamesob commented:
Okay, I think the new HEAD here is a pretty good compromise. Thanks for pushing me to avoid the duplication. 

Basically, I'm just reusing the `04_install.sh` script from within a container.

MarcoFalke commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+# See test/lint/README.md for usage.
```
nit: Given that this isn't used by the ci, mabye move this file to `test/lint` and remove this comment? (The paths below would need to be adjusted)

jamesob replied:
Yeah I started this way, but because the docker build context either (i) pulls in entire tree when run from root (super slow) and (ii) doesn't allow "back references" to files outside of the context, it kind of has to live here. There wasn't another good workaround that I could find.

MarcoFalke replied:
nit: Ah ok, maybe add a comment why this needs to live here.

MarcoFalke commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY ./docker-entrypoint.sh /entrypoint.sh
+COPY ./04_install.sh /install.sh
+
+RUN apt-get update && \
```
nit: apt update is already done by install.sh?

MarcoFalke commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+RUN apt-get update && \
+  /install.sh && \
+  mv /tmp/shellcheck-*/shellcheck /usr/bin/ && \
```
nit: Maybe just put the bin in this dir in `./ci/lint/04_install.sh` to avoid duplication here?

jamesob replied:
Good call, will do.

MarcoFalke commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+  mv /tmp/shellcheck-*/shellcheck /usr/bin/ && \
+  echo 'alias lint="./ci/lint/06_script.sh"' >> ~/.bashrc && \
+  chmod 755 /entrypoint.sh && \
+  rm -rf /var/lib/apt/lists/*
```
(unrelated: It may be possible to remove the packages used to build python, but that seems overkill)

maflcko reviewed (approved):
lgtm. Left some nits (feel free to ignore)

jamesob commented:
Okay, so I've addressed @MarcoFalke's feedback. Because of the merge of https://github.com/bitcoin/bitcoin/commit/01ec5308bf616740c804247043046b23b483df5c, I've had to introduce kind of a controversial change that hopefully won't be at all controversial when we merge #26226 - because the `.python-version` file is difficult to access during container build for previously discussed reasons, I'm just basing this lint container on the preexisting `python:3.7-buster` image. One benefit here is that it's way quicker to build than compiling Python from source - but obviously it isn't Python 3.6 because 3.6 is so old that official Docker channels don't even offer an image for it.

I think this is totally okay for running linters locally, and once we bump to 3.7 (hopefully soon) this shouldn't raise any objections...

MarcoFalke commented:
3.6 was removed three days ago: https://github.com/docker-library/python/commit/2e0f1dc3300acc08b82a95857fd9c87ac1f91672

But yeah, seems fine to use any python version locally, as this should rarely lead to errors and the local thing is best-effort anyway.

aureleoules reviewed (commented):
From Cirrus:

> +# This is used by the 04_install.sh script; we can't read the Python version from 
> ^---- failure generated from lint-whitespace.py

Nice to see the linter linting the linter.
But when I execute the linter locally on this PR with `docker run --rm -v $(pwd):/bitcoin -w /bitcoin -it bitcoin-linter` I don't get the warning, why?

<details>
<summary>Details</summary>

```
docker run --rm -v $(pwd):/bitcoin -w /bitcoin -it bitcoin-linter
src/crypto/ctaes in HEAD currently refers to tree 1b6c31139a71f80245c09597c343936a8e41d021
src/crypto/ctaes in HEAD was last updated in commit 8501bedd7508ac514385806e191aec21ee978891 (tree 1b6c31139a71f80245c09597c343936a8e41d021)
GOOD
src/secp256k1 in HEAD currently refers to tree afb14c4f45fa7eb8651daced71faddebe0fba179
src/secp256k1 in HEAD was last updated in commit 9d47e7b71b2805430e8c7b43816efd225a6ccd8c (tree afb14c4f45fa7eb8651daced71faddebe0fba179)
GOOD
src/minisketch in HEAD currently refers to tree a749309e1b4edf48091edb57da6de2e8a16498a5
src/minisketch in HEAD was last updated in commit e9f1d8c27261070d209a28570ad36fe91531cdd2 (tree a749309e1b4edf48091edb57da6de2e8a16498a5)
GOOD
src/leveldb in HEAD currently refers to tree bd49d3c60051c1a8d8f030fa4ba7a29237fe7479
src/leveldb in HEAD was last updated in commit 1a463c70a368fb20316e03efac273438cf47baa3 (tree bd49d3c60051c1a8d8f030fa4ba7a29237fe7479)
GOOD
src/crc32c in HEAD currently refers to tree 5a1b5f7188bd2181e2a71d431fff282d50058e46
src/crc32c in HEAD was last updated in commit 08269e54a9a74e06c9fb72720a216d8c4d4532a2 (tree 5a1b5f7188bd2181e2a71d431fff282d50058e46)
GOOD
Args used        : 208
Args documented  : 219
Args undocumented: 0
set()
Args unknown     : 11
{'-zmqpubhashtx', '-zmqpubrawtx', '-zmqpubrawblockhwm', '-zmqpubhashblockhwm', '-zmqpubsequence', '-zmqpubhashblock', '-zmqpubhashtxhwm', '-zmqpubrawtxhwm', '-zmqpubrawblock', '-zmqpubsequencehwm', '-includeconf'}
Success: no issues found in 258 source files
```

</details>

jamesob commented:
> But when I execute the linter locally on this PR with docker run --rm -v $(pwd):/bitcoin -w /bitcoin -it bitcoin-linter I don't get the warning, why?

I'm working on this right now - it turns out that the lint script uses an envvar set by Cirrus CI to specify a range of commits to look at for whitespace issues. Another discrepancy that this PR will soon resolve I hope.

jamesob committed:
lint: specify the right commit range when running locally

When running lints on Cirrus, a special envvar is set ($CIRRUS_PR);
emulate this when running linters locally by setting $LOCAL_BRANCH
to any value.

jamesob commented:
> I'm working on this right now - it turns out that the lint script uses an envvar set by Cirrus CI to specify a range of commits to look at for whitespace issues. Another discrepancy that this PR will soon resolve I hope.

@aureleoules this is now fixed. I think this is pretty near ready to go from a code standpoint.

aureleoules reviewed (approved):
ACK b68e5a7feff3e93027e75da0cd9a590fef99aac1

I verified it builds and runs successfully and is now able to detect whitespace issues.

jamesob commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+COPY ./04_install.sh /install.sh
+
+RUN /install.sh && \
+  echo 'alias lint="./ci/lint/06_script.sh"' >> ~/.bashrc && \
```
Useful when dropping into bash (i.e. not using default command). Going to leave as-is.

stickies-v commented:
Concept ACK, thanks for working on this tooling

jamesob commented:
Is good? Pretty low-risk change that would be nice to have in.

stickies-v commented on test/lint/README.md:
```diff
@@ -1,5 +1,23 @@
+cd ./ci/lint
+docker build -t bitcoin-linter .
+
+cd /root/of/bitcoin/repo
```
since we're already assuming starting in the root 2 lines up, I'd just replace this with

```suggestion
cd ../..
```

stickies-v commented on ci/lint/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# See test/lint/README.md for usage.
+#
+# This container basically has to live in this directory in order to pull in the CI
+# install scripts. If it lived in the root directory, it would have to pull in the
+# entire repo as docker context during build; if it lived elsewhere, it wouldn't be
```
> If it lived in the root directory, it would have to pull in the entire repo as docker context during build;

Would you have preferred it lived in root? I think it makes more sense in `/ci/lint`, but if it would be better you could use `.dockerignore` to prevent adding everything to your docker context, as in this diff:

<details>
<summary>git diff</summary>

```diff
diff --git a/.dockerignore b/.dockerignore
new file mode 100644
index 000000000..142df6af9
--- /dev/null
+++ b/.dockerignore
@@ -0,0 +1,2 @@
+*
+!/ci/lint
\ No newline at end of file
diff --git a/ci/lint/Dockerfile b/Dockerfile
similarity index 91%
rename from ci/lint/Dockerfile
rename to Dockerfile
index 03c20c728..87c9bb85a 100644
--- a/ci/lint/Dockerfile
+++ b/Dockerfile
@@ -16,8 +16,8 @@ ENV LC_ALL=C.UTF-8
 ENV SKIP_PYTHON_INSTALL=1
 
 # Must be built from ./ci/lint/ for these paths to work.
-COPY ./docker-entrypoint.sh /entrypoint.sh
-COPY ./04_install.sh /install.sh
+COPY ./ci/lint/docker-entrypoint.sh /entrypoint.sh
+COPY ./ci/lint/04_install.sh /install.sh
 
 RUN /install.sh && \
   echo 'alias lint="./ci/lint/06_script.sh"' >> ~/.bashrc && \
```
</details>

MarcoFalke replied:
For reference, you can avoid passing the full context by using buildkit or podman. See

* https://docs.docker.com/build/buildkit/#overview
* https://docs.podman.io/en/latest/markdown/podman-build.1.html#name

stickies-v reviewed (approved):
ACK b68e5a7fe

Not an expert in our linter or Docker, but the code looks safe and reasonable to me. I tested locally, and verified that `bitcoin-linter` errors as expected after added a whitespace commit.

Looks low-risk to add and can be improved upon later if needed.

john-moffett reviewed (approved):
ACK b68e5a7feff3e93027e75da0cd9a590fef99aac1

Works as expected on Intel Mac and detected my test issues. Thanks!
